src/qibo/tensorflow/gates.py

```diff
@@ -211,17 +211,38 @@ def construct_unitary(self) -> tf.Tensor:
         return tf.linalg.diag(diag)
 
 
-class ZPow(TensorflowGate, base_gates.ZPow):
+class U1(TensorflowGate, base_gates.U1):
 
     def __init__(self, q, theta):
-        base_gates.ZPow.__init__(self, q, theta)
+        base_gates.U1.__init__(self, q, theta)
         TensorflowGate.__init__(self)
 
     def construct_unitary(self) -> tf.Tensor:
         t = tf.cast(self.parameter, dtype=DTYPES.get('DTYPECPX'))
         phase = tf.exp(1j * t)
-        diag = tf.concat([1, phase], axis=0)
-        return tf.linalg.diag(diag)
+        return tf.linalg.diag([1, phase])
+
+
+class U2(TensorflowGate, base_gates.U2):
+
+    def __init__(self, q, phi, lam):
+        base_gates.U2.__init__(self, q, phi, lam)
+        TensorflowGate.__init__(self)
+
+    def construct_unitary(self):
+        from qibo.tensorflow import cgates
+        return cgates.U2.construct_unitary(self)
+
+
+class U3(TensorflowGate, base_gates.U3):
+
+    def __init__(self, q, theta, phi, lam):
+        base_gates.U3.__init__(self, q, theta, phi, lam)
+        TensorflowGate.__init__(self)
+
+    def construct_unitary(self):
+        from qibo.tensorflow import cgates
+        return cgates.U3.construct_unitary(self)
```

Comment:
Ok, here and in similar places I have a question/problem. I was trying to put Qibo inside a `tf.py_function` so I could use it inside a `tf.function` pipeline, as expected this is possible when using some tf backend different from custom, however if we have numpy calls in tensorflow/gates.py the py_function will crash, so I may suggest to have a "redundancy" and repeat the instructions using pure tf primitives.

Reply:
That's a good point, however could you find a `tf.function` script that actually crashes? For example I just tried:
```Python
@tf.function
def apply(state):
    gate = gates.U3(0, theta=0.1, phi=0.2, lam=0.3)
    return gate(state)

state = tf.cast(np.random.random(nqubits * (2,)), dtype=tf.complex128)
fstate = apply(state)
```
and it works both with and without `@tf.function`. Also `circuit.compile()` seems to work as well, for example:
```Python
c = Circuit(1)
c.add(gates.U3(0, theta=0.1, phi=0.2, lam=0.3))
c.compile()
fstate = c()
```
Regardless of that, it is probably better if all `construct_unitary` for Tensorflow gates return `tf.Tensor` instead of numpy arrays for consistency. This can be done either with repeating code (as you say) or just with casting `tf.cast(cgates.U3.construct_unitary(self), dtype=DTYPES.get('DTYPECPX'))`. Let me know what you think.

Reply:
Right, I think these functions are fine because we return a np.array. I think we should keep the initialization full numpy, to reduce the overhead that we have observed with tf primitives. I will open a PR if I get the `py_function` error again.